pkg/reconcile/reconcile.go

```diff
@@ -81,13 +83,20 @@ type Reconciler interface {
 	// Reconciler performs a full reconciliation for the object referred to by the Request.
 	// The Controller will requeue the Request to be processed again if an error is non-nil or
 	// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-	Reconcile(Request) (Result, error)
+	Reconcile(Context, Request) (Result, error)
 }
 
 // Func is a function that implements the reconcile interface.
-type Func func(Request) (Result, error)
+type Func func(Context, Request) (Result, error)
 
 var _ Reconciler = Func(nil)
 
 // Reconcile implements Reconciler.
-func (r Func) Reconcile(o Request) (Result, error) { return r(o) }
+func (r Func) Reconcile(ctx Context, o Request) (Result, error) { return r(ctx, o) }
+
+// Context is the context type passed to each Reconcile function, it MUST satisfy the context.Context interface.
+type Context struct {
```

Comment:
Why not an interface embedding `context.Context` and exposing `Logger() logr.Logger` as a function?
That would allow way more flexibility in passing arbitrary implementation structs which can be richer in functionality once casted to the right type, or more detailed interface.

Reply:
The main reason to prefer a struct was to avoid `ctx.Logger().Info()` everywhere, it's mostly for readability but I can see that an interface could be nice.

Reply:
Actually, if we go with what @alvaroaleman suggested below, creating a `context` package an interface would be great.

We could have `context.Background()` or similar things that would use the global logger, and other functions like `context.WithLogger()` that would take a context and change the internal logger implementation.

Reply:
Yes :tada:!

Reply:
> The main reason to prefer a struct was to avoid ctx.Logger().Info() everywhere, it's mostly for readability but I can see that an interface could be nice.

I don't personally think that's too much of an issue, as the consumer can always do `ctxLog := ctx.Logger()` if they get tired of calling the logger function all the time.